File: 백준/Gold/17220. 마약수사대/마약수사대.js
let input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

let [n, m] = input[0].split(" ").map(Number);
let map = Array.from({ length: n }, () => []);
let root = Array.from({ length: n }).fill(0);

for (let i = 0; i < m; i++) {
  let [start, end] = input[i + 1]
    .split(" ")
    .map((v) => v.charCodeAt(0) - "A".charCodeAt(0));
  map[start].push(end);
  root[end]++;
}

let cnt = 0;
let visit = Array.from({ length: n }).fill(false);

let [num, ...targetList] = input[m + 1].split(" ");
num = +num;
targetList = targetList.map((v) => v.charCodeAt(0) - "A".charCodeAt(0));

for (let target of targetList) {
  visit[target] = true;
}

const dfs = (start) => {
  if (!visit[start]) {
    let list = map[start];

    for (let t of list) {
      if (!visit[t]) {
        cnt++;
        dfs(t);
      }
    }
  }
};

for (let i = 0; i < n; i++) {
  if (root[i] === 0) {
    dfs(i);
  }
}

console.log(cnt);

File: 백준/Gold/17220. 마약수사대/마약수사대.java
import java.io.*;
import java.util.*;

public class Main {
    static int cnt;
    static boolean[] visit;
    static List<List<Integer>> graph;

    static void dfs(int start) {
        if (!visit[start]) {
            List<Integer> target = graph.get(start);
            for (int t : target) {
                if (!visit[t]) {
                    cnt++;
                    dfs(t);
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);
        graph = new ArrayList<>();
        visit = new boolean[n];
        int[] root = new int[n];

        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            input = br.readLine().split(" ");
            graph.get(input[0].charAt(0) - 'A').add(input[1].charAt(0) - 'A');
            root[input[1].charAt(0) - 'A']++;
        }

        input = br.readLine().split(" ");
        int arrest = Integer.parseInt(input[0]);

        for (int i = 0; i < arrest; i++) {
            int target = input[i + 1].charAt(0) - 'A';
            visit[target] = true;
        }

        for (int i = 0; i < n; i++) {
            if (root[i] == 0) {
                dfs(i);
            }
        }

        System.out.println(cnt);
    }
}
